source/common/lua/lua.h

```diff
@@ -0,0 +1,378 @@
+   * should not be used after some event. This allows us to mark objects as dead so that if they
+   * are used again they will throw a Lua error and not reach our code.
+   * @param state supplies the calling LuaState.
+   */
+  int checkDead(lua_State* state) {
+    if (dead_) {
+      return luaL_error(state, "object used outside of proper scope");
+    }
+    return 0;
+  }
+
+  /**
+   * Mark an object as dead so that a checkDead() call will throw an error. See checkDead().
+   */
+  void markDead() {
+    dead_ = true;
+    ENVOY_LOG(trace, "marking dead {} at {}", typeid(T).name(), static_cast<void*>(this));
+    onMarkDead();
+  }
+
+  /**
+   * Mark an object as live so that a checkDead() call will not throw an error. See checkDead().
+   */
+  void markLive() {
+    dead_ = false;
+    ENVOY_LOG(trace, "marking live {} at {}", typeid(T).name(), static_cast<void*>(this));
+    onMarkLive();
+  }
+
+protected:
+  /**
+   * Called from markDead() when an object is marked dead. This is effectively a C++ destructor for
+   * Lua/C objects. Objects can perform inline cleanup or mark other objects as dead if needed. It
+   * can also be used to protect objects from use if they get assigned to a global variable and
+   * used across coroutines.
+   */
+  virtual void onMarkDead() {}
+
+  /**
+   * Called from markLive() when an object is marked live. This is a companion to onMarkDead(). See
+   * the comments there.
+   */
+  virtual void onMarkLive() {}
+
+private:
+  bool dead_{};
+};
+
+/**
+ * This is basically a Lua smart pointer. The idea is that given a Lua object, if we want to
+ * guarantee that Lua won't destroy it, we need to reference it. This wraps the reference
+ * functionality. While a LuaRef owns an object it's guaranteed that Lua will not GC it.
+ */
+template <typename T> class LuaRef {
+public:
+  /**
+   * Create an empty LuaRef.
+   */
+  LuaRef() { reset(); }
+
+  /**
+   * Create a LuaRef from an object.
+   * @param object supplies the object. Generally this is the return value from a Object::create()
+   *        call. The object must be at the top of the Lua stack.
+   * @param leave_on_stack supplies whether to leave the object on the stack or not when the ref
+   *        is constructed.
+   */
+  LuaRef(const std::pair<T*, lua_State*>& object, bool leave_on_stack) {
+    reset(object, leave_on_stack);
+  }
+
+  ~LuaRef() { unref(); }
+  T* get() { return object_.first; }
+
+  /**
+   * Same as the LuaRef non-default constructor, but post-construction.
+   */
+  void reset(const std::pair<T*, lua_State*>& object, bool leave_on_stack) {
+    unref();
+
+    if (leave_on_stack) {
+      lua_pushvalue(object.second, -1);
+    }
+
+    object_ = object;
+    ref_ = luaL_ref(object_.second, LUA_REGISTRYINDEX);
+    ASSERT(ref_ != LUA_REFNIL);
+  }
+
+  /**
+   * Return a LuaRef to its default/empty state.
+   */
+  void reset() {
```

Comment:
This overloading of `reset()` is a bit confusing; in one we take a ref, the other we release?!

Reply:
This is the same API as unique_ptr, etc. so I thought it would be obvious. If you would prefer a different name that's fine with me. Let me know.